File: test/ts-stub-result-overlay/src/custom-containers.ts
import { IBaboonServiceRt } from "./testpkg/pkg0/baboon-service-rt";

export type Result<S, E> =
    | { readonly tag: "Success"; readonly value: S }
    | { readonly tag: "Failure"; readonly error: E };

export const ResultServiceRt: IBaboonServiceRt = {
    pure<L, R>(value: R): Result<R, L> {
        return { tag: "Success", value };
    },
    fail<L, R>(error: L): Result<R, L> {
        return { tag: "Failure", error };
    },
    leftMap<A, B, C>(value: Result<B, A>, f: (a: A) => C): Result<B, C> {
        return value.tag === "Failure"
            ? { tag: "Failure", error: f(value.error) }
            : { tag: "Success", value: value.value };
    },
    flatMap<A, B, C>(value: Result<B, A>, f: (b: B) => Result<C, A>): Result<C, A> {
        return value.tag === "Failure"
            ? { tag: "Failure", error: value.error }
            : f(value.value);
    },
};

File: test/rs-stub-result-overlay/src/custom_containers.rs
use crate::testpkg::pkg0::baboon_service_rt::IBaboonServiceRt;

pub enum MyResult<S, E> {
    Success(S),
    Failure(E),
}

pub struct ResultServiceRt;

impl IBaboonServiceRt for ResultServiceRt {
    fn pure<L: 'static, R>(&self, value: R) -> MyResult<R, L> {
        MyResult::Success(value)
    }
    fn fail<L: 'static, R>(&self, error: L) -> MyResult<R, L> {
        MyResult::Failure(error)
    }
    fn left_map<A: 'static, B, C: 'static, F: FnOnce(A) -> C>(&self, value: MyResult<B, A>, f: F) -> MyResult<B, C> {
        match value {
            MyResult::Failure(a) => MyResult::Failure(f(a)),
            MyResult::Success(b) => MyResult::Success(b),
        }
    }
    fn flat_map<A: 'static, B, C, F: FnOnce(B) -> MyResult<C, A>>(&self, value: MyResult<B, A>, f: F) -> MyResult<C, A> {
        match value {
            MyResult::Failure(a) => MyResult::Failure(a),
            MyResult::Success(b) => f(b),
        }
    }
}
